### ISSUE TITLE
[RFC] Context aware operations

### ISSUE THREAD

**Issue by BoShurik** [mark-gerarts/automapper-plus#30]
Hello!

I have the following use case: https://gist.github.com/BoShurik/7aafc932e9dcf74be2196a09d6da15bd

To build children elements I need access to parent object
```php
$config
    ->registerMapping(ItemDTO::class, Item::class)
    ->forMember('value', Operation::mapFrom(function (ItemDTO $dto) {
        return new Value(/*OrderDTO::$currency*/, $dto->value);
    }))
;
```

I suggest to use context:
```php
$config
    ->registerMapping(OrderDTO::class, Order::class)
    ->forMember('value', Operation::mapFrom(function (OrderDTO $dto) {
        return new Value($dto->currency, $dto->value);
    }))
    ->forMember('items', Operation::mapFromWithMapper(function (OrderDTO $dto, AutoMapper $mapper) {
        $items = [];
        foreach ($dto->items as $item) {
            $items[] = $mapper->map($item, Item::class, [
               'currency' => $dto->currency,
            ]);
        }

        return $items;
    }))
;
$config
    ->registerMapping(ItemDTO::class, Item::class)
    ->forMember('value', Operation::mapFromWithContext(function (ItemDTO $dto, array $context) {
        return new Value($context['currency'] ?? 'EUR', $dto->value);
    }))
;
```

Not sure, but may be it solves problem described in #16

**Comment by mark-gerarts** [comment]
Hello @BoShurik. I like this suggestion, it seems really useful. I've skimmed through the code a bit, and it certainly seems possible to implement. However, to be consistent I think I would want to pass the context to every operation (where it makes sense), and not make a specific `mapFromWithContext`. This way you could just do something like this:

```php
$config->registerMapping(ItemDTO::class, Item::class)
    ->forMember('value', Operation::mapFrom(function (ItemDTO $dto, AutoMapper $mapper, array $context) {
        // ...
    });
```

I'm gonna give this some more tought, and I hope to start experimenting a bit next week. I'll get back to you!

As a side note, I plan on making the automapper always available to the `MapFrom` operation, instead of using a separate `mapFromWithMapper` operation. Seems a bit more user-friendly to me.

**Comment by BoShurik** [comment]
> As a side note, I plan on making the automapper always available to the MapFrom operation, instead of using a separate mapFromWithMapper operation. Seems a bit more user-friendly to me.

It was my next suggestion :)

**Comment by mark-gerarts** [comment]
Hi @BoShurik. The feature has been implemented on the `feature/context` branch. You can view the documentation [here](https://github.com/mark-gerarts/automapper-plus/tree/feature/context#adding-context). If you have any remarks or suggestions, let me know. Otherwise I'm merging it into master and tagging a new release.

**Comment by BoShurik** [comment]
Looks good to me

**Comment by mark-gerarts** [comment]
Released in 1.3.0! :tada:
